src/extensions/scratch3_makeymakey/index.js

```diff
@@ -0,0 +1,372 @@
+     * hat block fires repeatedly.
+     * @param {object} args - the block arguments.
+     * @property {number} KEY - a key code.
+     * @param {object} util - utility object provided by the runtime.
+     */
+    whenMakeyKeyPressed (args, util) {
+        let key = args.KEY;
+        // Convert the key arg, if it is a KEY_ID, to the key name used by
+        // the Keyboard io module.
+        if (SCRATCH_KEY_NAME[args.KEY]) {
+            key = SCRATCH_KEY_NAME[args.KEY];
+        }
+        const isDown = util.ioQuery('keyboard', 'getKeyIsDown', [key]);
+        return (isDown && this.frameToggle);
+    }
+
+    /*
+     * A function called on the KEY_PRESSED event, to update the key press
+     * buffer and check if any of the key sequences have been completed.
+     * @param {string} key A scratch key name.
+     */
+    keyPressed (key) {
+        // Store only the first word of the Scratch key name, so that e.g. when
+        // "left arrow" is pressed, we store "LEFT", which matches KEY_ID_LEFT
+        key = key.split(' ')[0];
+        key = key.toUpperCase();
+        this.keyPressBuffer.push(key);
+        // Keep the buffer under the length limit
+        if (this.keyPressBuffer.length > KEY_BUFFER_LENGTH) {
+            this.keyPressBuffer.shift();
+        }
+        // Check the buffer for each sequence in use
+        for (const str in this.sequences) {
+            const arr = this.sequences[str].array;
+            // Bail out if we don't have enough presses for this sequence
+            if (this.keyPressBuffer.length < arr.length) {
+                continue;
+            }
+            let missFlag = false;
+            // Slice the buffer to the length of the sequence we're checking
+            const bufferSegment = this.keyPressBuffer.slice(-1 * arr.length);
+            for (let i = 0; i < arr.length; i++) {
+                if (arr[i] !== bufferSegment[i]) {
+                    missFlag = true;
+                }
+            }
+            // If the miss flag is false, the sequence matched the buffer
+            if (!missFlag) {
+                this.sequences[str].completed = true;
+                // Clear the completed flag after a timeout. This is necessary because
+                // the hat is edge-triggered (not event triggered). Multiple hats
+                // may be checking the same sequence, so this timeout gives them enough
+                // time to all trigger before resetting the flag.
+                setTimeout(() => {
+                    this.sequences[str].completed = false;
+                }, SEQUENCE_HAT_TIMEOUT);
+            }
+        }
+    }
+
+    /*
+     * Add a key sequence to the set currently being checked on each key press.
+     * @param {string} sequenceString a string of space-separated KEY_IDs.
+     * @param {array} sequenceArray an array of KEY_IDs.
+     */
+    addSequence (sequenceString, sequenceArray) {
```

Comment:
Why doesn't it matter that there is `addSequence` but not `removeSequence`? What happens after you don't listen for a sequence anymore? What if someone put a variable into the hat block and then changed that variable in a forever loop?

Reply:
I thought about it (only a little bit, really) and decided that it would be a rare and weird situation that is (I think) no more harmful than other things that you can do, likely only on purpose, to mess up Scratch, such as storing tons of data in variables and lists. Then again, we should find a way to fix it! I'll make a follow-up issue to consider different options.